File: demos/node/probe.mjs
#!/usr/bin/env node
// demos 能力探针 —— Node.js / JavaScript 实现。协议见 demos/SPEC.md。
//
// 模式：默认跑全部；--echo-child；--read-stdin；--argv-probe A B
import { execFileSync, spawnSync } from "node:child_process";
import { createHash } from "node:crypto";
import { gzipSync, gunzipSync } from "node:zlib";
import { createServer, createConnection } from "node:net";
import fs from "node:fs";
import path from "node:path";
import { Worker } from "node:worker_threads";

const SELF = path.resolve(process.argv[1]);
const NODE = process.execPath;

// 递归护栏（见 python 实现里的同款说明）：探针自我调用，模式分派写错就会指数派生。
const DEPTH = Number(process.env.DEMO_DEPTH || "0");
const MAX_DEPTH = 3;

const UNICODE_S = "中文-日本語-한국어-🚀";
const UNICODE_CHARS = 12;
const UNICODE_BYTES = 31;
const LARGE_BYTES = 262144;
const SHA256_ABC = "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad";

const results = [];
function record(id, status, detail = "") {
  const clean = String(detail).replace(/[\t\r\n]/g, " ").slice(0, 160);
  results.push([id, status, clean]);
  console.log(`CAP ${id} ${status} ${clean}`);
}

async function run(id, fn) {
  try {
    const detail = await fn();
    record(id, "PASS", detail || "");
  } catch (err) {
    record(id, "FAIL", `${err.name}: ${err.message}`);
  }
}

function spawn(args, stdinData) {
  const env = { ...process.env, DEMO_DEPTH: String(DEPTH + 1) };
  return execFileSync(NODE, [SELF, ...args], {
    input: stdinData === undefined ? "" : stdinData,
    env,
    timeout: 60000,
    encoding: "utf8",
    stdio: ["pipe", "pipe", "pipe"],
  });
}

// stdout 与 stderr 需要分开捕获，所以 A03/A17 单独走 spawnSync
function spawnSplit(args) {
  const env = { ...process.env, DEMO_DEPTH: String(DEPTH + 1) };
  return spawnSync(NODE, [SELF, ...args], { env, encoding: "utf8", timeout: 60000 });
}

// ---------------------------------------------------------------- 子进程模式
const argv = process.argv.slice(2);
if (argv[0] === "--echo-child") {
  process.stdout.write("child-ok\n");
  process.stderr.write("child-err\n");
  process.exit(0);
}
if (argv[0] === "--read-stdin") {
  let buf = "";
  process.stdin.setEncoding("utf8");
  process.stdin.on("data", (c) => { buf += c; });
  process.stdin.on("end", () => {
    const line = buf.split("\n")[0].trim();
    process.stdout.write(line === "ping" ? "pong\n" : `unexpected:${line}\n`);
    process.exit(0);
  });
} else if (argv[0] === "--argv-probe") {
  const rest = argv.slice(1);
  process.stdout.write(`argv-probe:${rest.length}:${rest.join(":")}\n`);
  process.exit(0);
} else {
  if (DEPTH > MAX_DEPTH) {
    console.error(`FATAL: probe recursion depth ${DEPTH} exceeded`);
    process.exit(3);
  }
  main().then((code) => process.exit(code));
}

async function main() {
  await run("A01_argv", () => {
    const got = spawn(["--argv-probe", "alpha", "beta"]).trim();
    if (got !== "argv-probe:2:alpha:beta") throw new Error(`argv echo ${got}`);
    return "2 args round-tripped";
  });

  await run("A02_env", () => {
    const v = process.env.DEMO_LANG_TAG;
    if (v !== "demos-capability") throw new Error(`DEMO_LANG_TAG=${v}`);
    return "env visible";
  });

  await run("A03_streams", () => {
    const r = spawnSplit(["--echo-child"]);
    if (r.stdout.trim() !== "child-ok") throw new Error(`stdout=${r.stdout}`);
    if (!r.stderr.includes("child-err")) throw new Error(`stderr=${r.stderr}`);
    if (r.stdout.includes("child-err")) throw new Error("stderr leaked into stdout");
    return "stdout/stderr separated";
  });

  await run("A04_stdin", () => {
    const got = spawn(["--read-stdin"], "ping\n").trim();
    if (got !== "pong") throw new Error(`reply=${got}`);
    return "ping->pong";
  });

  await run("A05_file_rw", () => {
    fs.writeFileSync("a05.txt", "hello-io", "utf8");
    const got = fs.readFileSync("a05.txt", "utf8");
    if (got !== "hello-io") throw new Error(`read back ${got}`);
    return "text round-trip ok";
  });

  await run("A06_file_append", () => {
    fs.writeFileSync("a06.txt", "a", "utf8");
    fs.appendFileSync("a06.txt", "b", "utf8");
    const got = fs.readFileSync("a06.txt", "utf8");
    if (got !== "ab") throw new Error(`append ${got}`);
    return "append ok";
  });

  await run("A07_file_binary", () => {
    const payload = Buffer.from(Array.from({ length: 256 }, (_, i) => i));
    fs.writeFileSync("a07.bin", payload);
    const got = fs.readFileSync("a07.bin");
    if (!got.equals(payload)) throw new Error(`binary mismatch len=${got.length}`);
    return "256 bytes incl 0x00/0xFF";
  });

  await run("A08_file_stat", () => {
    fs.writeFileSync("a08.txt", "statted", "utf8");
    if (!fs.existsSync("a08.txt")) throw new Error("not created");
    const st = fs.statSync("a08.txt");
    if (st.size !== 7) throw new Error(`size=${st.size}`);
    fs.unlinkSync("a08.txt");
    if (fs.existsSync("a08.txt")) throw new Error("unlink failed");
    return "size=7 then removed";
  });

  await run("A09_dir_ops", () => {
    fs.rmSync("a09dir", { recursive: true, force: true });
    fs.mkdirSync("a09dir");
    fs.writeFileSync(path.join("a09dir", "inner.txt"), "x", "utf8");
    const entries = fs.readdirSync("a09dir");
    if (!entries.includes("inner.txt")) throw new Error(`listing=${entries}`);
    fs.unlinkSync(path.join("a09dir", "inner.txt"));
    fs.rmdirSync("a09dir");
    if (fs.existsSync("a09dir")) throw new Error("rmdir failed");
    return "mkdir/list/rmdir ok";
  });

  await run("A10_temp_file", () => {
    const dir = fs.mkdtempSync(path.join(process.cwd(), "a10-"));
    const f = path.join(dir, "t.tmp");
    fs.writeFileSync(f, "temp-content", "utf8");
    const got = fs.readFileSync(f, "utf8");
    if (got !== "temp-content") throw new Error(`temp read ${got}`);
    fs.rmSync(dir, { recursive: true, force: true });
    return "unique temp file";
  });

  await run("A11_unicode", () => {
    if ([...UNICODE_S].length !== UNICODE_CHARS) throw new Error("codepoint count");
    const bytes = Buffer.byteLength(UNICODE_S, "utf8");
    if (bytes !== UNICODE_BYTES) throw new Error(`bytes=${bytes}`);
    fs.writeFileSync("a11.txt", UNICODE_S, "utf8");
    const got = fs.readFileSync("a11.txt", "utf8");
    if (got !== UNICODE_S) throw new Error("round-trip mismatch");
    return `${UNICODE_CHARS} codepoints / ${UNICODE_BYTES} bytes`;
  });

  await run("A12_large_io", () => {
    const payload = Buffer.alloc(LARGE_BYTES, 0x61);
    fs.writeFileSync("a12.bin", payload);
    const got = fs.readFileSync("a12.bin");
    if (got.length !== LARGE_BYTES) throw new Error(`len=${got.length}`);
    return "262144 bytes ok";
  });

  await run("A13_time", async () => {
    const nowMs = Date.now();
    if (nowMs < 1577836800000) throw new Error(`epoch=${nowMs}`);
    const t0 = process.hrtime.bigint();
    await new Promise((r) => setTimeout(r, 50));
    const deltaMs = Number(process.hrtime.bigint() - t0) / 1e6;
    if (deltaMs < 40) throw new Error(`sleep 只测到 ${deltaMs.toFixed(1)} ms`);
    return `sleep ${deltaMs.toFixed(0)}ms`;
  });

  await run("A14_random", () => {
    const v = Math.floor(Math.random() * 1000);
    if (!(v >= 0 && v < 1000)) throw new Error(`out of range ${v}`);
    return "in [0,1000)";
  });

  await run("A15_container", () => {
    const data = [5, 3, 9, 1, 7, 3].sort((a, b) => a - b);
    if (data.join(",") !== "1,3,3,5,7,9") throw new Error(`sorted=${data}`);
    const m = new Map([["a", 1]]);
    m.set("b", 2);
    if (m.get("a") !== 1 || m.get("b") !== 2 || m.size !== 2) throw new Error("map");
    return "sort + map ok";
  });

  await run("A16_error", () => {
    let caught = false;
    try {
      fs.readFileSync("definitely-missing-file-xyz", "utf8");
    } catch {
      caught = true;
    }
    if (!caught) throw new Error("no error raised");
    return "exception caught";
  });

  await run("A17_subprocess", () => {
    const r = spawnSplit(["--echo-child"]);
    if (r.status !== 0) throw new Error(`child exit=${r.status}`);
    if (r.stdout.trim() !== "child-ok") throw new Error(`child stdout=${r.stdout}`);
    return "child exit=0, stdout captured";
  });

  await run("A18_concurrency", async () => {
    const sab = new SharedArrayBuffer(4);
    const view = new Int32Array(sab);
    const code = `
      const { workerData } = require('node:worker_threads');
      const v = new Int32Array(workerData);
      for (let i = 0; i < 1000; i++) Atomics.add(v, 0, 1);
    `;
    const mk = () => new Promise((res, rej) => {
      const w = new Worker(code, { eval: true, workerData: sab });
      w.on("exit", res);
      w.on("error", rej);
    });
    await Promise.all([mk(), mk()]);
    if (Atomics.load(view, 0) !== 2000) throw new Error(`total=${Atomics.load(view, 0)}`);
    return "2 workers -> 2000";
  });

  await run("A19_tcp_loopback", async () => {
    return await new Promise((resolve, reject) => {
      const srv = createServer((conn) => {
        conn.on("data", (d) => {
          if (d.toString() !== "tcp-ping") reject(new Error(`recv=${d}`));
          conn.end("tcp-pong");
        });
      });
      srv.on("error", reject);
      srv.listen(0, "127.0.0.1", () => {
        const port = srv.address().port;
        const cli = createConnection({ host: "127.0.0.1", port }, () => cli.write("tcp-ping"));
        let reply = "";
        cli.setEncoding("utf8");
        cli.on("data", (d) => { reply += d; });
        cli.on("end", () => {
          srv.close();
          if (reply !== "tcp-pong") reject(new Error(`reply=${reply}`));
          else resolve("loopback send/recv ok");
        });
        cli.on("error", reject);
      });
    });
  });

  await run("B01_sha256", () => {
    const got = createHash("sha256").update("abc").digest("hex");
    if (got !== SHA256_ABC) throw new Error(`sha256=${got}`);
    return "sha256(abc) ok";
  });

  await run("B02_base64", () => {
    const enc = Buffer.from("abc").toString("base64");
    if (enc !== "YWJj") throw new Error(`b64=${enc}`);
    if (Buffer.from(enc, "base64").toString("utf8") !== "abc") throw new Error("decode");
    return "encode+decode ok";
  });

  await run("B03_json", () => {
    const obj = { k: [1, 2, 3], n: "v" };
    const text = JSON.stringify(obj);
    const back = JSON.parse(text);
    if (back.k.join(",") !== "1,2,3" || back.n !== "v") throw new Error("round-trip");
    return "serialize+parse ok";
  });

  await run("B04_regex", () => {
    if (!/^[a-z]+-[0-9]{3}$/.test("abc-123")) throw new Error("positive match failed");
    if (/^[a-z]+-[0-9]{3}$/.test("ABC-123")) throw new Error("negative match unexpected");
    return "match+reject ok";
  });

  await run("B05_gzip", () => {
    const payload = Buffer.from("gzip-payload-".repeat(8));
    if (!gunzipSync(gzipSync(payload)).equals(payload)) throw new Error("gzip round-trip");
    return "compress+decompress ok";
  });

  const counts = { PASS: 0, FAIL: 0, SKIP: 0 };
  for (const [, status] of results) counts[status]++;
  console.log(`SUMMARY node ${counts.PASS} ${counts.FAIL} ${counts.SKIP}`);
  return counts.FAIL ? 1 : 0;
}

File: demos/java/Probe.java
// demos 能力探针 —— Java 实现。协议见 demos/SPEC.md。
//
// 模式（照抄 SPEC §4，跨语言必须同名）：
//   默认              跑全部检查
//   --echo-child      往 stdout 写 child-ok、往 stderr 写 child-err，exit 0
//   --read-stdin      读一行 stdin，内容为 ping 时回 pong
//   --argv-probe A B  回显 argv（argv-probe:2:A:B）
import java.io.BufferedReader;
import java.io.File;
import java.io.FileDescriptor;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.regex.Pattern;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;

public final class Probe {

    private static final String UNICODE_S = "中文-日本語-한국어-🚀";
    private static final int UNICODE_CODEPOINTS = 12;
    private static final int UNICODE_BYTES = 31;
    private static final int LARGE_BYTES = 262144;
    private static final String SHA256_ABC =
            "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad";
    private static final int MAX_DEPTH = 3;

    private static int depth;
    private static int npass;
    private static int nfail;
    private static int nskip;

    private interface Check {
        String run() throws Exception;
    }

    private static final class SkipCheck extends Exception {
        private static final long serialVersionUID = 1L;

        SkipCheck(String reason) {
            super(reason);
        }
    }

    private static void record(String cid, String status, String detail) {
        String clean = detail == null ? "" : detail.replace('\t', ' ').replace('\r', ' ').replace('\n', ' ');
        if (clean.length() > 160) {
            clean = clean.substring(0, 160);
        }
        if ("PASS".equals(status)) {
            npass++;
        } else if ("FAIL".equals(status)) {
            nfail++;
        } else {
            nskip++;
        }
        System.out.print("CAP " + cid + " " + status + " " + clean + "\n");
        System.out.flush();
    }

    private static void run(String cid, Check c) {
        try {
            record(cid, "PASS", c.run());
        } catch (SkipCheck e) {
            record(cid, "SKIP", e.getMessage());
        } catch (Exception e) {
            record(cid, "FAIL", e.getClass().getSimpleName() + ": " + e.getMessage());
        }
    }

    // ------------------------------------------------------------- 自我调用
    private static String javaExe() {
        String home = System.getProperty("java.home");
        File f = new File(home, "bin/java");
        if (!f.exists()) {
            f = new File(home, "bin/java.exe");
        }
        return f.exists() ? f.getAbsolutePath() : "java";
    }

    /** 本类被加载的路径（即 javac -d 的输出目录）。子进程用同一个 classpath 启动。 */
    private static String selfClasspath() {
        try {
            URL u = Probe.class.getProtectionDomain().getCodeSource().getLocation();
            return Paths.get(u.toURI()).toString();
        } catch (Exception e) {
            return ".";
        }
    }

    private static final class Result {
        final int code;
        final String out;
        final String err;

        Result(int code, String out, String err) {
            this.code = code;
            this.out = out;
            this.err = err;
        }
    }

    private static Result spawn(List<String> args, String stdinData) throws Exception {
        List<String> cmd = new ArrayList<>();
        cmd.add(javaExe());
        cmd.add("-Dfile.encoding=UTF-8");
        cmd.add("-cp");
        cmd.add(selfClasspath());
        cmd.add("Probe");
        cmd.addAll(args);

        ProcessBuilder pb = new ProcessBuilder(cmd);
        pb.environment().put("DEMO_DEPTH", String.valueOf(depth + 1));
        Process p = pb.start();
        if (stdinData != null) {
            OutputStream os = p.getOutputStream();
            os.write(stdinData.getBytes(StandardCharsets.UTF_8));
            os.flush();
            os.close();
        } else {
            p.getOutputStream().close();
        }

        StringBuilder so = new StringBuilder();
        StringBuilder se = new StringBuilder();
        Thread t1 = new Thread(() -> drain(p.getInputStream(), so));
        Thread t2 = new Thread(() -> drain(p.getErrorStream(), se));
        t1.start();
        t2.start();
        if (!p.waitFor(60, java.util.concurrent.TimeUnit.SECONDS)) {
            p.destroyForcibly();
            throw new IOException("child timed out");
        }
        t1.join(5000);
        t2.join(5000);
        return new Result(p.exitValue(), so.toString(), se.toString());
    }

    private static void drain(java.io.InputStream in, StringBuilder sb) {
        try (BufferedReader r = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
            String line;
            while ((line = r.readLine()) != null) {
                sb.append(line).append('\n');
            }
        } catch (IOException e) {
            // 子进程退出时管道可能被关闭，属正常
        }
    }

    // ------------------------------------------------------------------ Tier A
    private static String a01() throws Exception {
        List<String> a = new ArrayList<>();
        a.add("--argv-probe");
        a.add("alpha");
        a.add("beta");
        Result r = spawn(a, null);
        String got = r.out.trim();
        if (!"argv-probe:2:alpha:beta".equals(got)) {
            throw new AssertionError("argv 回显不符: " + got);
        }
        return "2 args round-tripped";
    }

    private static String a02() {
        String v = System.getenv("DEMO_LANG_TAG");
        if (!"demos-capability".equals(v)) {
            throw new AssertionError("DEMO_LANG_TAG=" + v);
        }
        return "env visible";
    }

    private static String a03() throws Exception {
        List<String> a = new ArrayList<>();
        a.add("--echo-child");
        Result r = spawn(a, null);
        if (!"child-ok".equals(r.out.trim())) {
            throw new AssertionError("stdout=" + r.out.trim());
        }
        if (!r.err.contains("child-err")) {
            throw new AssertionError("stderr=" + r.err.trim());
        }
        if (r.out.contains("child-err")) {
            throw new AssertionError("stderr leaked into stdout");
        }
        return "stdout/stderr separated";
    }

    private static String a04() throws Exception {
        List<String> a = new ArrayList<>();
        a.add("--read-stdin");
        Result r = spawn(a, "ping\n");
        if (!"pong".equals(r.out.trim())) {
            throw new AssertionError("reply=" + r.out.trim());
        }
        return "ping->pong";
    }

    private static String a05() throws Exception {
        Files.write(Paths.get("a05.txt"), "hello-io".getBytes(StandardCharsets.UTF_8));
        String got = new String(Files.readAllBytes(Paths.get("a05.txt")), StandardCharsets.UTF_8);
        if (!"hello-io".equals(got)) {
            throw new AssertionError("read back " + got);
        }
        return "text round-trip ok";
    }

    private static String a06() throws Exception {
        Path p = Paths.get("a06.txt");
        Files.write(p, "a".getBytes(StandardCharsets.UTF_8));
        java.nio.file.StandardOpenOption app = java.nio.file.StandardOpenOption.APPEND;
        Files.write(p, "b".getBytes(StandardCharsets.UTF_8), app);
        String got = new String(Files.readAllBytes(p), StandardCharsets.UTF_8);
        if (!"ab".equals(got)) {
            throw new AssertionError("append 结果 " + got);
        }
        return "append ok";
    }

    private static String a07() throws Exception {
        byte[] payload = new byte[256];
        for (int i = 0; i < 256; i++) {
            payload[i] = (byte) i;
        }
        Files.write(Paths.get("a07.bin"), payload);
        byte[] got = Files.readAllBytes(Paths.get("a07.bin"));
        if (!Arrays.equals(payload, got)) {
            throw new AssertionError("二进制往返不一致 len=" + got.length);
        }
        return "256 bytes incl 0x00/0xFF";
    }

    private static String a08() throws Exception {
        Path p = Paths.get("a08.txt");
        Files.write(p, "statted".getBytes(StandardCharsets.UTF_8));
        if (!Files.exists(p)) {
            throw new AssertionError("file not created");
        }
        long size = Files.size(p);
        if (size != 7) {
            throw new AssertionError("size=" + size);
        }
        Files.delete(p);
        if (Files.exists(p)) {
            throw new AssertionError("delete failed");
        }
        return "size=7 then removed";
    }

    private static String a09() throws Exception {
        Path d = Paths.get("a09dir");
        if (Files.exists(d)) {
            Files.delete(d.resolve("inner.txt"));
            Files.delete(d);
        }
        Files.createDirectory(d);
        Files.write(d.resolve("inner.txt"), "x".getBytes(StandardCharsets.UTF_8));
        boolean found = false;
        File[] entries = d.toFile().listFiles();
        if (entries == null) {
            throw new AssertionError("listFiles returned null");
        }
        for (File e : entries) {
            if ("inner.txt".equals(e.getName())) {
                found = true;
            }
        }
        if (!found) {
            throw new AssertionError("listing missing inner.txt");
        }
        Files.delete(d.resolve("inner.txt"));
        Files.delete(d);
        if (Files.exists(d)) {
            throw new AssertionError("rmdir failed");
        }
        return "mkdir/list/rmdir ok";
    }

    private static String a10() throws Exception {
        Path p = Files.createTempFile(Paths.get("."), "demos-", ".tmp");
        try {
            Files.write(p, "temp-content".getBytes(StandardCharsets.UTF_8));
            String got = new String(Files.readAllBytes(p), StandardCharsets.UTF_8);
            if (!"temp-content".equals(got)) {
                throw new AssertionError("temp read " + got);
            }
        } finally {
            Files.deleteIfExists(p);
        }
        return "unique temp file";
    }

    private static String a11() throws Exception {
        int cps = UNICODE_S.codePointCount(0, UNICODE_S.length());
        if (cps != UNICODE_CODEPOINTS) {
            throw new AssertionError("codepoints=" + cps);
        }
        byte[] raw = UNICODE_S.getBytes(StandardCharsets.UTF_8);
        if (raw.length != UNICODE_BYTES) {
            throw new AssertionError("bytes=" + raw.length);
        }
        Path p = Paths.get("a11.txt");
        Files.write(p, raw);
        String got = new String(Files.readAllBytes(p), StandardCharsets.UTF_8);
        if (!UNICODE_S.equals(got)) {
            throw new AssertionError("round-trip mismatch");
        }
        return UNICODE_CODEPOINTS + " codepoints / " + UNICODE_BYTES + " bytes";
    }

    private static String a12() throws Exception {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < LARGE_BYTES / 8; i++) {
            sb.append("abcdefgh");
        }
        byte[] payload = sb.toString().getBytes(StandardCharsets.US_ASCII);
        if (payload.length != LARGE_BYTES) {
            throw new AssertionError("payload len=" + payload.length);
        }
        Files.write(Paths.get("a12.bin"), payload);
        byte[] got = Files.readAllBytes(Paths.get("a12.bin"));
        if (!Arrays.equals(payload, got)) {
            throw new AssertionError("len=" + got.length);
        }
        return LARGE_BYTES + " bytes ok";
    }

    private static String a13() throws Exception {
        long ms = System.currentTimeMillis();
        if (ms < 1577836800000L) {
            throw new AssertionError("epoch=" + ms);
        }
        long t0 = System.nanoTime();
        Thread.sleep(50);
        double delta = (System.nanoTime() - t0) / 1e6;
        if (delta < 40) {
            throw new AssertionError(String.format("sleep 只测到 %.1f ms", delta));
        }
        return String.format("sleep %.0fms", delta);
    }

    private static String a14() {
        int v = new java.util.Random().nextInt(1000);
        if (v < 0 || v >= 1000) {
            throw new AssertionError("out of range " + v);
        }
        return "in [0,1000)";
    }

    private static String a15() {
        int[] data = {5, 3, 9, 1, 7, 3};
        Arrays.sort(data);
        if (!Arrays.toString(data).equals("[1, 3, 3, 5, 7, 9]")) {
            throw new AssertionError("sorted=" + Arrays.toString(data));
        }
        Map<String, Integer> m = new HashMap<>();
        m.put("a", 1);
        m.put("b", 2);
        if (m.get("a") != 1 || m.get("b") != 2 || m.size() != 2) {
            throw new AssertionError("map broken");
        }
        return "sort + map ok";
    }

    private static String a16() {
        boolean caught = false;
        try {
            Files.readAllBytes(Paths.get("definitely-missing-file-xyz"));
        } catch (IOException e) {
            caught = true;
        }
        if (!caught) {
            throw new AssertionError("no error raised for missing file");
        }
        return "IOException caught";
    }

    private static String a17() throws Exception {
        List<String> a = new ArrayList<>();
        a.add("--echo-child");
        Result r = spawn(a, null);
        if (r.code != 0) {
            throw new AssertionError("child exit=" + r.code);
        }
        if (!"child-ok".equals(r.out.trim())) {
            throw new AssertionError("child stdout=" + r.out.trim());
        }
        return "child exit=0, stdout captured";
    }

    private static String a18() throws Exception {
        AtomicInteger total = new AtomicInteger(0);
        Thread t1 = new Thread(() -> {
            for (int i = 0; i < 1000; i++) {
                total.incrementAndGet();
            }
        });
        Thread t2 = new Thread(() -> {
            for (int i = 0; i < 1000; i++) {
                total.incrementAndGet();
            }
        });
        t1.start();
        t2.start();
        t1.join();
        t2.join();
        if (total.get() != 2000) {
            throw new AssertionError("total=" + total.get());
        }
        return "2 threads -> 2000";
    }

    private static String a19() throws Exception {
        InetAddress loop = InetAddress.getByName("127.0.0.1");
        final ServerSocket srv = new ServerSocket(0, 4, loop);
        final int port = srv.getLocalPort();
        final String[] recv = new String[1];

        Thread t = new Thread(() -> {
            try (Socket conn = srv.accept()) {
                byte[] buf = new byte[64];
                int n = conn.getInputStream().read(buf);
                recv[0] = new String(buf, 0, n, StandardCharsets.US_ASCII);
                conn.getOutputStream().write("tcp-pong".getBytes(StandardCharsets.US_ASCII));
                conn.getOutputStream().flush();
            } catch (IOException e) {
                recv[0] = "ERR:" + e.getMessage();
            }
        });
        t.start();

        String got;
        try (Socket cli = new Socket(loop, port)) {
            cli.getOutputStream().write("tcp-ping".getBytes(StandardCharsets.US_ASCII));
            cli.getOutputStream().flush();
            byte[] buf = new byte[64];
            int n = cli.getInputStream().read(buf);
            got = new String(buf, 0, n, StandardCharsets.US_ASCII);
        }
        t.join(10000);
        srv.close();
        if (!"tcp-ping".equals(recv[0]) || !"tcp-pong".equals(got)) {
            throw new AssertionError("recv=" + recv[0] + " reply=" + got);
        }
        return "loopback send/recv ok";
    }

    // ------------------------------------------------------------------ Tier B
    private static String b01() throws Exception {
        byte[] d = MessageDigest.getInstance("SHA-256").digest("abc".getBytes(StandardCharsets.UTF_8));
        StringBuilder sb = new StringBuilder();
        for (byte b : d) {
            sb.append(String.format("%02x", b));
        }
        if (!SHA256_ABC.equals(sb.toString())) {
            throw new AssertionError("sha256=" + sb);
        }
        return "sha256(abc) ok";
    }

    private static String b02() {
        String enc = Base64.getEncoder().encodeToString("abc".getBytes(StandardCharsets.US_ASCII));
        if (!"YWJj".equals(enc)) {
            throw new AssertionError("b64=" + enc);
        }
        String dec = new String(Base64.getDecoder().decode(enc), StandardCharsets.US_ASCII);
        if (!"abc".equals(dec)) {
            throw new AssertionError("b64 decode mismatch");
        }
        return "encode+decode ok";
    }

    private static String b03() throws Exception {
        // JDK 标准库**没有** JSON：java.util 与 java.text 都不含序列化器。
        throw new SkipCheck("JDK 标准库无 JSON（需 Gson/Jackson 等第三方库）");
    }

    private static String b04() {
        Pattern p = Pattern.compile("^[a-z]+-[0-9]{3}$");
        if (!p.matcher("abc-123").matches()) {
            throw new AssertionError("positive match failed");
        }
        if (p.matcher("ABC-123").matches()) {
            throw new AssertionError("negative match unexpected");
        }
        return "match+reject ok";
    }

    private static String b05() throws Exception {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 8; i++) {
            sb.append("gzip-payload-");
        }
        byte[] payload = sb.toString().getBytes(StandardCharsets.US_ASCII);

        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        try (GZIPOutputStream gz = new GZIPOutputStream(bos)) {
            gz.write(payload);
        }
        byte[] packed = bos.toByteArray();

        ByteArrayOutputStream back = new ByteArrayOutputStream();
        try (GZIPInputStream gz = new GZIPInputStream(new ByteArrayInputStream(packed))) {
            byte[] buf = new byte[512];
            int n;
            while ((n = gz.read(buf)) > 0) {
                back.write(buf, 0, n);
            }
        }
        if (!Arrays.equals(payload, back.toByteArray())) {
            throw new AssertionError("gzip round-trip mismatch");
        }
        return "compress+decompress ok";
    }

    // -------------------------------------------------------------------- 主流程
    public static void main(String[] argv) throws Exception {
        // ⚠️ 先把 stdout/stderr 换成显式 UTF-8 的 PrintStream。
        // JEP 400 之后 `-Dfile.encoding` 只影响 Charset.defaultCharset()，**管不到**
        // System.out：重定向到文件时它用的是原生编码（本机 CP936/GBK），
        // 于是中文 detail 被写成 GBK 字节 ⇒ 输出文件不是合法 UTF-8 ⇒
        // 下游 grep 判成 "binary file" 并丢掉整行（实测丢掉了带中文的 B03）。
        // 这里直接按字节钉死 UTF-8，彻底不依赖平台编码。
        System.setOut(new PrintStream(new FileOutputStream(FileDescriptor.out), true, StandardCharsets.UTF_8));
        System.setErr(new PrintStream(new FileOutputStream(FileDescriptor.err), true, StandardCharsets.UTF_8));

        String envDepth = System.getenv("DEMO_DEPTH");
        depth = envDepth == null ? 0 : Integer.parseInt(envDepth.trim());
        if (depth > MAX_DEPTH) {
            System.err.print("FATAL: probe recursion depth " + depth + " exceeded\n");
            System.exit(3);
        }

        // ⚠️ 模式分派必须先于任何检查执行：探针会自我调用，一旦派错就会指数级派生进程。
        if (argv.length >= 1) {
            if ("--echo-child".equals(argv[0])) {
                System.out.print("child-ok\n");
                System.err.print("child-err\n");
                System.out.flush();
                System.err.flush();
                System.exit(0);
            }
            if ("--read-stdin".equals(argv[0])) {
                BufferedReader r = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
                String line = r.readLine();
                line = line == null ? "" : line.trim();
                System.out.print("ping".equals(line) ? "pong\n" : "unexpected:" + line + "\n");
                System.out.flush();
                System.exit(0);
            }
            if ("--argv-probe".equals(argv[0])) {
                List<String> rest = new ArrayList<>();
                for (int i = 1; i < argv.length; i++) {
                    rest.add(argv[i]);
                }
                System.out.print("argv-probe:" + rest.size() + ":" + String.join(":", rest) + "\n");
                System.out.flush();
                System.exit(0);
            }
        }

        String[] ids = {
            "A01_argv", "A02_env", "A03_streams", "A04_stdin", "A05_file_rw", "A06_file_append",
            "A07_file_binary", "A08_file_stat", "A09_dir_ops", "A10_temp_file", "A11_unicode",
            "A12_large_io", "A13_time", "A14_random", "A15_container", "A16_error",
            "A17_subprocess", "A18_concurrency", "A19_tcp_loopback",
            "B01_sha256", "B02_base64", "B03_json", "B04_regex", "B05_gzip"
        };
        Check[] fns = {
            Probe::a01, Probe::a02, Probe::a03, Probe::a04, Probe::a05, Probe::a06,
            Probe::a07, Probe::a08, Probe::a09, Probe::a10, Probe::a11, Probe::a12,
            Probe::a13, Probe::a14, Probe::a15, Probe::a16, Probe::a17, Probe::a18, Probe::a19,
            Probe::b01, Probe::b02, Probe::b03, Probe::b04, Probe::b05
        };

        for (int i = 0; i < fns.length; i++) {
            run(ids[i], fns[i]);
        }

        System.out.print("SUMMARY java " + npass + " " + nfail + " " + nskip + "\n");
        System.out.flush();
        System.exit(nfail > 0 ? 1 : 0);
    }
}
